### ISSUE TITLE
Could not load the SqlServerDatabaseProvider DbProviderFactory.

### ISSUE THREAD

**Issue by ghost** [CollaboratingPlatypus/PetaPoco#470]
Version: PetaPoco.Compiled v6.0.317-beta
App type: dotnet core 2.0 Console Application

I am trying to create a new Database object in a console application, and am getting the error

    Could not load the SqlServerDatabaseProvider DbProviderFactory.

I downloaded the source, wrote a simple test, and it was blowing up on PetaPoco/Core/DatabaseProvider.cs line 158. The assemblyQualifiedNames list is actually a single hardcoded value for the assembly 

    "System.Data.SqlClient.SqlClientFactory, System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"

I took a guess and installed the nuget package for System.Data.SqlClient which is what fixed the issue. I haven't worked with DbProviderFactories long enough to intuitively understand what was missing, so this seems like an obscure message that took way to long to understand the real issue.

I am requesting two things:

1. Update the wiki page "wiki/Connection-string-provider-names" to list this assembly reference in a new column for each provider, and a quick note about required assemblies are listed in the column
2. Update the error message to be more descriptive. Maybe including the assemblies that were searched to something like 
    * "Could not load the XZY DbProviderFactory. Unable to find the Provider in System.Data.SqlClient".

**Comment by MarkLeMerise** [comment]
For anyone still wondering, this issue still exists for .NET Core 3+ and PetaPoco.Compiled v6.0.444-beta. Adding a Nuget package reference to `System.Data.SqlClient` got everything working again.

**Comment by asherber** [comment]
I'm not sure I would call this an "issue", although I agree that the docs could be improved. PetaPoco sits on top of the ADO.NET data providers and assumes that you have that layer in place. `System.Data.SqlClient` is built into the .NET Framework, which is why everything just works there, but you do need to grab the NuGet package for .NET Core. 

If you work with any of the other supported databases, like Postgres or MySql, those data providers need to be installed via NuGet for both Core and Framework apps.

**Comment by groogiam** [comment]
Adding System.Data.SqlClient nuget package does not seem to fix this issue. I have added this nuget package and still get this error with v6.0.446-beta

**Comment by asherber** [comment]
@groogiam  Can you post something minimal to show what you’re doing?

**Comment by groogiam** [comment]
[PetaPoco87.zip](https://github.com/CollaboratingPlatypus/PetaPoco/files/4257659/PetaPoco87.zip)

**Comment by asherber** [comment]
@groogiam Your sample project is missing the reference to the  `System.Data.SqlClient` nuget package. When I add this reference, the project works just fine for me.
```xml
  <ItemGroup>
    <PackageReference Include="PetaPoco.Compiled" Version="6.0.446-beta" />
    <PackageReference Include="System.Data.SqlClient" Version="4.8.0" />
  </ItemGroup>
```

**Comment by groogiam** [comment]
Whoops forgot to add that. Your right this does work fine. The place I originally tried this was in LinqPad 6 and it does not work there. See attached file. I'm guessing this an issue with LinqPad then.


[PetaPocoTest.linq.txt](https://github.com/CollaboratingPlatypus/PetaPoco/files/4258022/PetaPocoTest.linq.txt)

**Comment by asherber** [comment]
Hmmm. I feel like LinqPad _should_ work, but I haven't tried it yet. I'll keep this in mind.

**Comment by groogiam** [comment]
I created a bug report with the LinqPad folks. Well see what they say. Thanks for your help.

**Comment by groogiam** [comment]
Updating the .NET SDK from 3.1.101 to 3.1.102 seems to fix the LinqPad issue.

**Comment by antichaosdb** [comment]
We just upgraded our Azure functions from v2 to v3 and hit this issue, but after adding a reference to System.Data.SqlClient we got null reference exceptions in PetaPoco.  This happens during the resolving of the Provider factor in DatabaseProvider line 104
`ft = Type.GetType(assemblyName);`
Seems like there has been a recent change in the Azure runtime which is causing Type.GetType to throw errors.
My workaround was to create a provider that returns the factory without reflection:
```    
public class RealSqlServerDatabaseProvider : SqlServerDatabaseProvider
    {
        public override DbProviderFactory GetFactory()
        {
            return SqlClientFactory.Instance;
        }
    }
```

**Comment by beyond8848** [comment]
seems it's not feasible in vs 2019.   i have added the below packagereference as blew in my project.
<ItemGroup>
    <PackageReference Include="PetaPoco.Compiled" Version="6.0.446-beta" />
    <PackageReference Include="System.Data.SqlClient" Version="4.8.0" />
  </ItemGroup>

**Comment by cbusjay** [comment]
Updating from .NET SDK 3.1.101 to 3.1.102 worked for me as well (as @groogiam states) until recently with the release of .NET SDK 3.1.201.  It now produces the same error as before "Could not load the SqlServerDatabaseProvider DbProviderFactory."

@antichaosdb 's work around worked for me so thanks for that!

**Comment by biapar** [comment]
> For anyone still wondering, this issue still exists for .NET Core 3+ and PetaPoco.Compiled v6.0.444-beta. Adding a Nuget package reference to `System.Data.SqlClient` got everything working again.

It works for me.

**Comment by AsifulNobel** [comment]
@antichaosdb Were you able to find any other solution? I am also upgrading my app from Azure Functions v2 to v3 and facing this issue. Your solution worked for me. But It seems hacky.

**Comment by asherber** [comment]
@groogiam Do you have a link for the bug report you filed in LinqPad? I'm finding this frustrating and would like to follow up.

Basically, the following code finds the type in a console app for .NET Core 3.1 or 5.0, but does not find the type in a Linqpad project targeting the same runtimes.

```csharp
var aqn = "System.Data.SqlClient.SqlClientFactory, System.Data";
var type = Type.GetType(aqn);
```

Now, I know that under .NET Core, the type actually resides in the System.Data.SqlClient assembly, not System.Data. But I'm curious as to why my console app can resolve the type anyway, but Linqpad cannot.

**Comment by groogiam** [comment]
@asherber I can't find any response from my bug submission to LinqPad in my inbox. Shortly after I submitted it started working with the next minor version of .NET core 3.1 as per my comment above. This has not really been an issue for me as long as I add System.Data.SqlClient package to the LinqPad query. Sorry I can't be of more help.

**Comment by asherber** [comment]
The fact that System.Data.SqlClient didn't work properly with Linqpad was confirmed as an [issue in Linqpad](https://forum.linqpad.net/discussion/2502/system-data-sqlclient-types-not-found-in-lp6/p1?new=1), fixed in v6.15.6-beta.

**Comment by reinux** [comment]
Same issue with F# Interactive.
